src/ZkFold/Base/Protocol/Protostar/Fold.hs

```diff
@@ -0,0 +1,179 @@
+
+        ck = oracle i
+
+        initialAccumulator :: Accumulator (Vector n a) a c (Map Natural a)
+        initialAccumulator = Accumulator (AccumulatorInstance (P.pure zero) [hcommit ck [zero :: Map Natural a]] [] initE zero) [zero]
+
+
+instanceProof
+    :: forall n c a
+    .  Arithmetic a
+    => KnownNat n
+    => Scale a a
+    => HomomorphicCommit a [Map Natural a] c
+    => a
+    -> RecursiveCircuit n a
+    -> SPS.Input a (RecursiveCircuit n a)
+    -> InstanceProofPair (Vector n a) c (Map Natural a)
+instanceProof ck rc i = InstanceProofPair i (NARKProof [hcommit ck [m]] [m])
+    where
+        m = SPS.prover @a rc M.empty i []
+
+foldN
+    :: forall n c a
+    .  Arithmetic a
+    => P.Eq c
+    => AdditiveGroup c
+    => Scale a c
+    => Scale a a
+    => KnownNat n
+    => RandomOracle a a
+    => RandomOracle c a
+    => RandomOracle [c] a
+    => HomomorphicCommit a [a] c
+    => HomomorphicCommit a [Map Natural a] c
+    => Natural
+    -> a
+    -> RecursiveCircuit n a
+    -> SPS.Input a (RecursiveCircuit n a)
+    -> [P.Bool]
+    -> Accumulator (Vector n a) a c (Map Natural a)
+    -> Acc.KeyScale a
+    -> FoldResult n c a
+foldN iter ck rc i verifierResults acc ks
+  | iterations rc == 0 = FoldResult i acc (and verifierResults) (Acc.decider (transform ck rc i :: FS_CM n c a) (ck, ks) acc)
+  | otherwise = let (output, newAcc, newVerifierResult, newKs) = foldStep ck rc i acc ks
+                 in foldN iter ck (rc {iterations = iterations rc -! 1}) output (newVerifierResult : verifierResults) newAcc newKs
+
+executeAc :: forall n a . KnownNat n => RecursiveCircuit n a -> Vector n a -> Vector n a
+executeAc (RecursiveCircuit _ rc) i = eval rc i
+
+foldStep
+    :: forall n c a
+    .  Arithmetic a
+    => P.Eq c
+    => AdditiveGroup c
+    => KnownNat n
+    => Scale a c
+    => Scale a a
+    => RandomOracle a a
+    => RandomOracle c a
+    => RandomOracle [c] a
+    => HomomorphicCommit a [Map Natural a] c
+    => HomomorphicCommit a [a] c
+    => a
+    -> RecursiveCircuit n a
+    -> SPS.Input a (RecursiveCircuit n a)
+    -> Accumulator (Vector n a) a c (Map Natural a)
+    -> Acc.KeyScale a
+    -> (SPS.Input a (RecursiveCircuit n a), Accumulator (Vector n a) a c (Map Natural a), P.Bool, Acc.KeyScale a)
+foldStep ck rc i acc (Acc.KeyScale alphaSum muSum) = (newInput, newAcc, verifierAccepts, Acc.KeyScale (alphaSum + alphaPows) (muSum + scale (6 :: Natural) alpha))
+    where
+        fs :: FS_CM n c a
+        fs = transform ck rc i
+
+        nark@(InstanceProofPair _ narkProof) = instanceProof ck rc i
+        (newAcc, accProof) = Acc.prover fs ck acc nark
+
+        alpha :: a
+        alpha = oracle (acc^.x, i, narkCommits narkProof, accProof)
+
+        alphaPows = sum $ P.take (P.length accProof) $ (alpha ^) <$> [1 :: Natural ..]
+
+        verifierAccepts = Acc.verifier @_ @_ @_ @(Map Natural a) @(FS_CM n c a) i (narkCommits narkProof) (acc^.x) (newAcc^.x) accProof
```

Comment:
The verifier should always accept the proof when it is constructed with the `prover` algorithm. These checks are redundant here.